extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -303,70 +250,130 @@ export class AmpA4A extends AMP.BaseElement {
-        // via #validateAdResponse_.  See GitHub issue
-        // https://github.com/ampproject/amphtml/issues/4187
-        return creative && this.maybeRenderAmpAd_(creative);
-      })
-      .catch(error => this.promiseErrorHandler_(error));
+        // This block returns the ad URL, if one is available.
+        /** @return {!Promise<?string>} */
+        .then(() => {
+          checkStillCurrent(promiseId);
+          return this.getAdUrl();
+        })
+    // This block returns the (possibly empty) response to the XHR request.
+    /** @return {!Promise<?Response>} */
+    .then(adUrl => {
+      checkStillCurrent(promiseId);
+      this.adUrl_ = adUrl;
+      return adUrl && this.sendXhrRequest_(adUrl);
+    })
+    // The following block returns either the response (as a {bytes, headers}
+    // object), or null if no response is available / response is empty.
+    /** @return {!Promise<?{bytes: !ArrayBuffer, headers: !Headers}>} */
+    .then(fetchResponse => {
+      checkStillCurrent(promiseId);
+      if (!fetchResponse || !fetchResponse.arrayBuffer) {
+        return null;
+      }
+      // Note: Resolving a .then inside a .then because we need to capture
+      // two fields of fetchResponse, one of which is, itself, a promise,
+      // and one of which isn't.  If we just return
+      // fetchResponse.arrayBuffer(), the next step in the chain will
+      // resolve it to a concrete value, but we'll lose track of
+      // fetchResponse.headers.
+      return fetchResponse.arrayBuffer().then(bytes => {
+        return {
+          bytes,
+          headers: fetchResponse.headers,
+        };
+      });
+    })
+    // This block returns the ad creative and signature, if available; null
+    // otherwise.
+    /**
+     * @return {!Promise<?AdResponseDef>}
+     */
+    .then(responseParts => {
+      checkStillCurrent(promiseId);
+      return responseParts && this.extractCreativeAndSignature(
+          responseParts.bytes, responseParts.headers);
+    })
+    // This block returns the ad creative if it exists and validates as AMP;
+    // null otherwise.
+    /** @return {!Promise<?ArrayBuffer>} */
+    .then(creativeParts => {
+      checkStillCurrent(promiseId);
+      if (!creativeParts || !creativeParts.signature) {
+        return Promise.resolve(null);
+      }
+
+      // Alias for validCreativePromise's resolve function. This will be called
+      // if any of the fetched keys successfully validate the creative.
+      let resolveValidation;
+      // Promise that will resolve upon any successful validation.
+      const validCreativePromise = new Promise(resolve => {
+        // @param {!function(?ArrayBuffer)} resolve
+        resolveValidation = resolve;
+      });
+      // Promise that will resolve to null after all keys have been checked.
+      // Will call resolveValidation if a successful validation does happen.
+      const allKeysCheckedPromise = Promise.all(
```

Comment:
I meant https://github.com/ampproject/amphtml/pull/5091#discussion_r79500983, but we can't.